pkg/txgenerator/common.go

```diff
@@ -0,0 +1,89 @@
+package txgenerator
+
+import (
+	"errors"
+	"math/big"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/types"
+)
+
+// TxType represents the type of transaction to create
+type TxType = byte
+
+// BaseTxParams contains common parameters for all transaction types
+type BaseTxParams struct {
+	// Nonce is the transaction nonce
+	Nonce uint64
+	// GasLimit is the gas limit for the transaction
+	GasLimit uint64
+	// ChainID is the chain ID for the transaction
+	ChainID *big.Int
+
+	// For legacy transactions (TxType = LegacyTxType):
+	// GasPrice is the gas price in wei
+	GasPrice *big.Int
+
+	// For EIP-1559 transactions (TxType = DynamicFeeTxType):
+	// MaxFeePerGas is the maximum fee per gas in wei
+	MaxFeePerGas *big.Int
+	// MaxPriorityFeePerGas is the maximum priority fee per gas in wei
+	MaxPriorityFeePerGas *big.Int
+}
+
+// createTokenTransaction is a helper function to create a token transaction (ERC20/ERC721/ERC1155)
+// with the appropriate transaction type (legacy or EIP-1559)
+func createTokenTransaction(
+	tokenAddress common.Address,
+	nonce uint64,
+	gasLimit uint64,
+	chainID *big.Int,
+	gasPrice *big.Int,
+	maxFeePerGas *big.Int,
+	maxPriorityFeePerGas *big.Int,
+	data []byte,
+) (*types.Transaction, error) {
+	// Determine transaction type based on provided fields
+	var txType TxType
```

Comment:
Could we add a new param `txType TxType` to this `createTokenTransaction` function and move all checks below to a new function `func validateAndDetermineTxType(params BaseTxParams) (TxType, error)`?

That way we could validate params and resolve the tx type before we get to  `createTokenTransaction` call in functions that are calling it?